File: app/cpp-backend/graph-preprocessing/assign_levels.cpp
#include "assign_levels.h"

#include <stdexcept>
#include <deque>
#include <tuple>
#include <iostream>

#include "../logging/boost_logging.hpp"

namespace graph_preprocessing {

void assignLevelsInGraph(GraphInterface& graph) {
    auto rootPointerList = graph.getRootList();
    if (rootPointerList.empty()) {
        throw std::runtime_error{"Assign levels error: graph must have at least one root"};
    }

    std::deque<std::pair<uint32_t, int>> q;
    std::deque<std::pair<uint32_t, int>> qPrim;
    for (const auto root : rootPointerList) {
        q.push_back({root, 0});
    }

    std::vector<bool> shouldInspect(graph.getVertexCount(), true);
    int64_t remainingInspectionsCount = shouldInspect.size();
    std::vector<uint32_t> liveIvCollection;
    liveIvCollection.reserve(graph.getVertexCount());
    for (size_t u=0; u<graph.getVertexCount(); ++u) {
        liveIvCollection.emplace_back(graph.NR(u).size());
    }

    #define _extractFront(_q, _u, _uLevel) (std::tie(_u, _uLevel) = _q.front(), _q.pop_front())
    while (!(q.empty() && qPrim.empty())) {
        uint32_t u;
        int uLevel;
        if (!q.empty()) _extractFront(q, u, uLevel);
        else {
            _extractFront(qPrim, u, uLevel);
            logging::log_warning(
                "Taken out " + std::to_string(u) + ", level = " + std::to_string(uLevel) +
                " from secondary (qPrim) queue when assigning levels."
            );
        }

        if (shouldInspect[u]) {
            shouldInspect[u] = false;
            graph.setLevelForVertex(u, uLevel);
            for (const auto v : graph.N(u)) {
                // std::cout << u << " -> " << v << "\n";
                if (--liveIvCollection[v] == 0) {
                    q.push_back({v, uLevel+1});
                } else {
                    qPrim.push_back({v, uLevel+1});
                }
            }
            
        }

        --remainingInspectionsCount;
        if (remainingInspectionsCount == 0) break;
    }

    #undef _extractFront
}

}

File: app/python-backend/src/hierachical_layout_prototype/assign_levels.py
from collections import deque as queue
from copy import deepcopy

from graph import Graph, Vertex


def assign_levels(G: Graph):
    if G.vertex_levels_computed:
        raise Exception("Vertex levels already computed")

    roots = G.roots
    if len(roots) == 0:
        raise Exception("The graph must have at least one root")

    Q = queue()
    Q_prim = queue()
    for u in roots:
        Q.append((u, 0))

    should_ignore = [False for _ in range(G.vertex_count)]
    live_I_v_collection = [deepcopy(G.N_reversed(u)) for u in range(G.vertex_count)]

    while not (len(Q) == 0 and len(Q_prim) == 0):
        if len(Q) > 0:
            u, level = Q.popleft()
        else:
            u, level = Q_prim.popleft()

        if not should_ignore[u]:
            should_ignore[u] = True
            G.set_level_for_v(vertex_index=u, level=level)
            for v in G.N(u):
                live_I_v_collection[v].remove(u)
                I_v = live_I_v_collection[v]
                if len(I_v) > 0:
                    Q_prim.append((v, level + 1))
                else:
                    Q.append((v, level + 1))
